### PR TITLE
VideoPlayer: Add missing override keywords

### PR DESCRIPTION
**Pull request by garbear**
Split out from #12238

### CODE REVIEW

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxBXA.h [hunk 1]
```diff
@@ -56,17 +56,19 @@ class CDVDDemuxBXA : public CDVDDemux
...
+
+  // implementation of CDVDDemux
+  virtual void Reset() override;
+  virtual void Abort() override;
+  virtual void Flush() override;
+  virtual DemuxPacket* Read() override;
   bool SeekTime(double time, bool backwards = false, double* startpts = NULL) override { return false; }
-  void SetSpeed(int iSpeed) {};
-  int GetStreamLength() { return (int)m_header.durationMs; }
+  virtual void SetSpeed(int iSpeed) override {}
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxBXA.h]: having SetSpeed as pure virtual does not make much sense if we need those kind of overrides in subclasses, right?

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxVobsub.h [hunk 2]
```diff
@@ -35,19 +35,20 @@ class CDVDDemuxVobsub : public CDVDDemux
...
 
-  virtual void          Reset();
-  virtual void          Abort() {};
-  virtual void          Flush();
-  virtual DemuxPacket*  Read();
+  bool Open(const std::string& filename, int source, const std::string& subfilename);
+
+  // implementation of CDVDDemux
+  virtual void          Reset() override;
+  virtual void          Abort() override {}
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxVobsub.h]: should be promoted to superclass

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxVobsub.h [hunk 3]
```diff
@@ -35,19 +35,20 @@ class CDVDDemuxVobsub : public CDVDDemux
...
-  virtual void          SetSpeed(int speed) {}
+  virtual void          SetSpeed(int speed) override {}
   virtual CDemuxStream* GetStream(int index) const override { return m_Streams[index]; }
   virtual std::vector<CDemuxStream*> GetStreams() const override;
   virtual int           GetNrOfStreams() const override { return m_Streams.size(); }
-  virtual int           GetStreamLength()    { return 0; }
-  virtual std::string   GetFileName()        { return m_Filename; }
-
-  bool                  Open(const std::string& filename, int source, const std::string& subfilename);
+  virtual int           GetStreamLength() override { return 0; }
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxVobsub.h]: also promote to superclass

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxVobsub.h [hunk 4]
```diff
@@ -35,19 +35,20 @@ class CDVDDemuxVobsub : public CDVDDemux
...
+  virtual std::string   GetFileName() override { return m_Filename; }
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxVobsub.h]: same here

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxBXA.h (hunk 1 above)
Comment (garbear) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxBXA.h]: You tell me :) You could make the argument that nearly any virtual method could have a trivial base class implementation. On the other hand, there might be value in being explicit by declaring the lack of any effect in its subclass.

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxVobsub.h (hunk 4 above)
Comment (garbear) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxVobsub.h]: but the superclass doesn't have access to the private `m_Filename`?

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxBXA.h (hunk 1 above)
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxBXA.h]: I don't kill warnings for the sake of killing warnings. We can take the opportunity to improve things that pop up here or just do nothing.
SetSpeed obvously was hacked into the interface at some time for a special case. Now we can take corrective action.

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxVobsub.h (hunk 4 above)
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxVobsub.h]: then it needs a closer look

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxBXA.h (hunk 1 above)
Comment (garbear) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxBXA.h]: ok, if you don't see anything obvious I'll look into these issues later tonight

Comment (garbear) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxBXA.h]: SetSpeed() doesn't look like a hack. It's called upon a `CDVDMsg::PLAYER_SETSPEED` msg. `DVDDemuxFFmpeg` uses this function to decide what kinds of frames should be discarded. `DVDDemuxFFmpeg` is the only subclass that implements this, so the superclass should provide the trivial implementation.

**Comment by garbear** [comment]
@FernetMenta added a commit moving all the functions you commented on (and more) to the superclass. Also added a commit for an unused function I just found.

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxBXA.h (hunk 1 above)
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxBXA.h]: SetSpeed for the ffmpeg demuxer has a single purpose: it pauses network streams that are handled directly by ffmpeg. this internal information is exposed to the upper layers and all other demuxers and devs how write demuxers have to deal with this. This is what I call a hack. disacrding frames on the demuxers is max nonsense, won't work this way. Onother hack.

Comment (garbear) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxBXA.h]: What's the next step? Drop this function?

**Comment by garbear** [comment]
Comment got hidden after the rebase, so I'll reply here

> SetSpeed for the ffmpeg demuxer has a single purpose: it pauses network streams that are handled directly by ffmpeg. this internal information is exposed to the upper layers and all other demuxers and devs how write demuxers have to deal with this. This is what I call a hack. disacrding frames on the demuxers is max nonsense, won't work this way. Onother hack.

What's the next step? Drop this function?

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemux.h [hunk 5]
```diff
@@ -243,29 +243,29 @@ class CDVDDemux
...
    * Flush the demuxer, if any data is kept in buffers, this should be freed now
    */
-  virtual void Flush() = 0;
+  virtual void Flush() { }
 
   /*
    * Read a packet, returns NULL on error
    *
    */
-  virtual DemuxPacket* Read() = 0;
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemux.h]: why did you change even the required methods? now the class is not pure virtual anymore.

Comment (garbear) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemux.h]: because function is not universally required: https://github.com/xbmc/xbmc/pull/12250/files#diff-3b3e940dfa4c42a41d871daaaabe3d70L37

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxBXA.h (hunk 1 above)
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxBXA.h]: demuxer could handle this internally. it knows about the nature of the stream and if it should be paused. if read is not called for a given time, it can pause ffmpeg. when player continues reading, the stream can continue as well. why should upper layers bother with those internals.

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemux.h (hunk 5 above)
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemux.h]: CDVDDemuxCC is not a regular demuxer. It was done before we got DemuxMultiSource and will be migrated.

**Comment by Rechi** [comment]
@garbear this needs a rebase

**Comment by garbear** [comment]
Just to make sure, the commit "VideoPlayer: fixup 2baba89" is a rebase error and should be dropped, right?

**Comment by FernetMenta** [comment]
@garbear yes, 2baba892693a68d32ad4fb54df7018a456e1d367 should not be in this PR here

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxBXA.h (hunk 1 above)
Comment (garbear) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxBXA.h]: The ownership of the input stream is confusing. It's held by both VideoPlayer and the demuxer. What's the reason for this? Would moving ownership to the demuxer exclusively solve this issue?

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemux.h (hunk 5 above)
Comment (garbear) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemux.h]: Which methods of the ones I changed are required?

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxBXA.h (hunk 1 above)
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxBXA.h]: lifetime of input stream is not bound to lifetime of demuxer. SetSpeed was hacked into the code violating a basic OO principle: information hiding

File: xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemux.h (hunk 5 above)
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemux.h]: all but SetSpeed

Comment (garbear) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemux.h]: but you commented that I should promote `Abort()`, `GetStreamLength()` and `GetFileName()` to the superclass?

Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemux.h]: right, I mean those are required by VP. if overridden they can be used to change behaviour.  VP uses those methods to control demuxers and get info from demuxers. SetSpeed does not belong to this category. it was implemented because the demuxer wanted info from VP.
